garak/generators/huggingface.py

```diff
@@ -50,53 +52,134 @@ def _set_hf_context_len(self, config):
+            params["device"] = self.device
+
+        args = {}
+
+        parameters = inspect.signature(hf_constructor).parameters
+
+        if "model" in parameters:
+            args["model"] = self.name
+            # expand for
+            parameters = {"do_sample": True} | parameters
+        else:
+            # callable is for a Pretrained class also map standard `pipeline` params
+            from transformers import pipeline
+
+            parameters = (
+                {"low_cpu_mem_usage": True}
+                | parameters
+                | inspect.signature(pipeline).parameters
+            )
+
+        for k in parameters:
+            if k == "model":
+                continue  # special case `model` comes from `name` in the generator
+            if k in params:
+                val = params[k]
+                if k == "torch_dtype" and hasattr(torch, val):
+                    args[k] = getattr(
+                        torch, val
+                    )  # some model type specific classes do not yet support direct string representation
+                    continue
+                if (
+                    k == "device"
+                    and "device_map" in parameters
+                    and "device_map" in params
+                ):
+                    # per transformers convention hold `device_map` before `device`
+                    continue
+                args[k] = params[k]
+
+        return args
+
+    def _select_hf_device(self):
+        """Determine the most efficient device for tensor load, hold any existing `device` already selected"""
+        import torch.cuda
+
+        selected_device = None
+        if self.hf_args["device"] is not None:
+            if isinstance(self.hf_args["device"], int):
+                # this assumes that indexed only devices selections means `cuda`
+                selected_device = torch.device("cuda:" + str(self.hf_args["device"]))
+            else:
+                selected_device = torch.device(self.hf_args["device"])
+
+        if selected_device is None:
+            selected_device = torch.device(
+                "cuda"
+                if torch.cuda.is_available()
+                else "mps" if torch.backends.mps.is_available() else "cpu"
+            )
+
+        if isinstance(selected_device, torch.device) and selected_device.type == "mps":
+            os.environ["PYTORCH_ENABLE_MPS_FALLBACK"] = "1"
+            logging.debug("Enabled MPS fallback environment variable")
+
+        logging.debug(
+            "Using %s, based on torch environment evaluation", selected_device
+        )
+        return selected_device
+
 
 class Pipeline(Generator, HFCompatible):
     """Get text generations from a locally-run Hugging Face pipeline"""
 
+    DEFAULT_PARAMS = Generator.DEFAULT_PARAMS | {
+        "generations": 10,
+        "hf_args": {
+            "torch_dtype": "float16",
+            "do_sample": True,
+            "device": None,
+        },
```

Comment:
It might be more intuitive to put this into HFCompatible, even though not all classes will use all parameters. I would like to have it possible to set e.g. `torch_dtype` in one place for all `huggingface` generators, so that when one switches from `Pipeline` to `Model` due to compatibility issues (which happens not-never), these options can be carried over.

Reply:
The current implementation already supports for this.

```
plugins:
  generators:
    huggingface:
      hf_args:
        torch_dtype: float16
      Model:
        hf_args:
            device: 0
```

The above will set `hf_args["torch_dtype"]` for all `huggingface` generators, and merge with any other `DEFAULT_PARAMS["hf_args"]` values, I am hesitant to place these in `HFCompatible` as I see added complexity in attempting to have a common set of `hf_args` for `transformers`.

Reply:
Wait, aren't we already in `HFCompatible` here?

I don't want an extra config node - I just am not convinced a different default torch_dtype setting (e.g.) should be made in each generator class. But actually if we only do it in `HFCompatible` then we should be good.

Reply:
We are in `Pipeline` which happens to also be composed with `HFCompatible`.

We could have `HFCompatible` implement a `DEFAULT_PARAMS` and merge it with the `Pipeline` and `Generator` defaults with something like 
``` Python
    DEFAULT_PARAMS = Generator.DEFAULT_PARAMS | HFCompatible.DEFAULT_PARAMS | {
        "generations": 10,
    }
```

I suspect this would quickly become error prone and get missed when creating new classes that need to be composed as with `HFCompatible`.

We could also consider changing the inheritance pattern to be `HFCompatible(Generator)` -> `Pipeline(HFCompatible)` and use 
``` Python
class HFCompatible(Generator):
    DEFAULT_PARAMS = Generator.DEFAULT_PARAMS | {
        "hf_args": {
            "torch_dtype": "float16",
            "do_sample": True,
            "device": None,
        }
    }

...

class Pipeline(HFCompatible):
    DEFAULT_PARAMS = HFCompatible.DEFAULT_PARAMS | {
        "generations": 10,
    }
```

Either pattern I have a concern about defaults that only make sense for a specific transformer slipping into the common set. As well as possible need to recursively merge `hf_args` if a new class needs a unique default that it cannot put in `HFCompatible.DEFAULT_PARAMS`.

I am a major proponent of code reuse, just not sure the trade off is worth it here. Happy to accommodate if this is a deal breaker.